Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs

```diff
@@ -86,39 +86,49 @@ public virtual void ProcessBlock(Block block)
 
             // If the new block previous hash is the same as the 
             // wallet hash then just pass the block to the manager. 
+            var newTip = this.chain.GetBlock(block.GetHash());
+
+            // new block is not in the main chain nothing more to do right now. 
+            if (newTip == null)
+                return; // reorg
+
             if (block.Header.HashPrevBlock != this.walletTip.HashBlock)
             {
-                // set a temporary wallet tip in case we are not on the best chain anymore.
-                var walletTip = this.WalletTip;
-
                 // If previous block does not match there might have 
                 // been a reorg, check if the wallet is still on the main chain.
                 ChainedBlock inBestChain = this.chain.GetBlock(this.walletTip.HashBlock);
                 if (inBestChain == null)
                 {
                     // The current wallet hash was not found on the main chain.
-                    // A reorg happenend so bring the wallet back top the last known fork.
+                    // A reorg happened so bring the wallet back top the last known fork.
 
-                    var fork = walletTip;
+                    var fork = this.walletTip;
 
                     // We walk back the chained block object to find the fork.
                     while (this.chain.GetBlock(fork.HashBlock) == null)
                         fork = fork.Previous;
 
                     this.walletManager.RemoveBlocks(fork);
 
-                    // update temporary wallet tip to be able to catch up again.
-                    walletTip = this.chain.GetBlock(fork.HashBlock);
+                    // update the wallet tip to be able to catch up again.
+                    this.walletTip = fork;
                 }
 
-                ChainedBlock incomingBlock = this.chain.GetBlock(block.GetHash());
-                if (incomingBlock.Height > walletTip.Height)
+                // The new tip can be ahead or behind the wallet
+                // If the new tip is ahead we try to bring the wallet up to the new tip.
+                // If the new tip is behind we just check the wallet and the tip ad in the same chain.
+            
+                if (newTip.Height > this.walletTip.Height)
                 {
+                    // check that the new tip is in the chain of the wallet tip
+                    var findTip = newTip.FindAncestorOrSelf(this.walletTip.HashBlock);
+                    Guard.Assert(findTip == this.walletTip); // this should never happen
+
```

Comment:
As per my analysis above, this is wrong for two reasons:
1) FindAncestorOrSelf should probably never be used in such circumstances as it can take a lot of CPU looping through the whole chain down to genesis.
2) It can certainly be the case that `findTip != walletTip` exactly because `newTip` and `fork` can be on completely different chains. More specifically, consider that `newTip` is not on best chain and `walletTip` is. In that case, this triggers the assert.

Reply:
So to further explain, on line https://github.com/dangershony/StratisBitcoinFullNode/blob/64df43a3641428632552610dcfb68134b653511e/Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs#L94 the chain changes to new chain -> `newTip` is not on the chain and the rest should be obvious.

Reply:
1. This will only happen on reorg so it should not happen that much.

>2. It can be the case that findTip != walletTip exactly because newTip and fork can be on different chains. 

Hmm good point (probably really rare but still valid).
I made the assumption that if we reorged then wallet and new tip are on the same chain.

Perhaps in this case the solution is to do nothing instead of Assert.

Reply:
I have very hard time understanding why `FindAncestorOrSelf` is implemented in a way it is. This just feels wrong from theoretical perspective. It is O(N) function and it should be O(1) using dictionary. As such, FAOS should never be used in case there is a chance that the hash you are looking for is not on the same chain. It should also not be used if the hash could be far from the block you are starting with. By far I mean anything above 50 items. 

So assert is wrong, yes, but just by removing it, the code is still unwell.

Reply:
Let's say I'm not sure we want to fix this issue right now, or using the design like this that is very complex to think about.

Reply:
There is an O(1) method but its in ConcurrentChain and we can not use ConcurrentChain.
You would want to use `FindAncestorOrSelf ` (or a variation of it) so not to depend on concurrent chain.
You can think of it in the same way as your initial session suggestion where a reorg will not happen in that used session, The selected ChainedBlock is representing the chain at that point in time and its guaranteed to not reorg.

Perhaps we could improve the look up on `FindAncestorOrSelf ` .

Reply:
If `findTip` is not found we can replace the assert with doing nothing giving executing back and letting the next round try to recover the wallet, this is a very very rare case of ace condition.

Reply:
Ah I see, this is to avoid using the chain ... In general, I would say FAOS is OK to use if you are absolutely sure that those two chained blocks are on the same chain (and not so far away, which is always true for minor reorgs). So can we possibly make that guarantee?

Reply:
We can't guarantee no, but I would say this would be a very rare case that 'findTip' is not in 'this.walltTip` for two reasons:
- Big reorgs are not so common.
- Earlier we checked that `findTip` is on the chain and we have passed the reorg code.

So both blocks by now should be on the chain, this check is to ensure that.
There would have multiple reorgs in the right moment for this to happen.
Also this is not happy path 99% of blocks will never go down this path.

Reply:
It is correct that this is rare and thus the inefficiency is not critical. I will try to reread the code as if the assert was replaced with IF and see if it makes any sense to me.

Reply:
OK, so I've read the code and this else block
```
                    // check that the new tip is in the chain of the wallet tip
                    var findTip = this.walletTip.FindAncestorOrSelf(newTip.HashBlock);
                    Guard.Assert(findTip == newTip); // this should never happen
```
should probably be also treated similarly - no guard and don't process block 
btw `var`, vars everywhere in this code :-)

Then 
```
 next = newTip.GetAncestor(next.Height + 1);
```
This will be mostly OK, but looks ugly from theoretical point of view. I'd replace it with better approach. Do you know where I am going here?

Then yes, the big while trying to somehow find and wait for a block if it is not there that's crazy, but we know that it is subject for replacement so OK.

Overall yes, it "should" work with guards replaced and having in the mind all the previous comments.

Reply:
OK awesome I will fix the `var` stuff and remove the asserts and we can review again.

Reply:
yup